kubeadm/create-cluster.sh

```diff
@@ -0,0 +1,150 @@
+gcloud compute firewall-rules create "$CLUSTER_NAME-default-external" --network "$CLUSTER_NAME" --allow "tcp:22,tcp:3389,tcp:6443" --source-ranges "0.0.0.0/0"
+
+# prepare control plane startup script
+SETUP_STARTUP_CP=$TMP_DIR/setup-startup-cp.sh
+cat setup.sh >> "$SETUP_STARTUP_CP"
+cat startup-cp.sh >> "$SETUP_STARTUP_CP"
+chmod +x "$SETUP_STARTUP_CP"
+
+# create CP instance and run the startup script on it
+# to debug the startup script use:
+#   sudo google_metadata_script_runner --script-type startup --debug
+# to view the startup logs see:
+#   /var/log/syslog
+CP_COMMAND=$(gcloud compute instances create "$CP_NODE_NAME" \
+	--metadata-from-file startup-script="$SETUP_STARTUP_CP" \
+	--image-family ubuntu-1804-lts \
+	--image-project ubuntu-os-cloud \
+	--machine-type $MACHINE_TYPE_CP \
+	--network "$CLUSTER_NAME" \
+	--labels=cluster="$CLUSTER_NAME" \
+	--tags kube-control-plane \
+	--format "json(networkInterfaces[0].networkIP, networkInterfaces[0].accessConfigs[0].natIP)")
+
+# wait for kubeadm init
+SECONDS=0
+set +o errexit
+while true; do
+	timeout 30 gcloud compute ssh "root@${CP_NODE_NAME}" --command "KUBECONFIG=/etc/kubernetes/admin.conf kubectl version"
+	# shellcheck disable=SC2181
+	if [ $? -eq 0 ]; then
+		gcloud compute scp "root@${CP_NODE_NAME}:/etc/kubernetes/admin.conf" "$KUBECONFIG"
+		break
+	else
+		if [ $SECONDS -ge $TIMEOUT_CP_READY ]; then
+			echo "Error: Failed waiting for control plane to become ready. Dumping its /var/log/syslog:"
+			timeout 30 gcloud compute ssh "root@${CP_NODE_NAME}" --command "sudo cat /var/log/syslog"
+			exit 1
+		fi
+		sleep 10
+	fi
+done
+set -o errexit
+
+# replace the private with the public IP in the kubeconfig
+PRIVATE_IP=$(echo "$CP_COMMAND" | jq -r '.[] | .networkInterfaces[0].networkIP')
+PUBLIC_IP=$(echo "$CP_COMMAND" | jq -r '.[] | .networkInterfaces[0].accessConfigs[0].natIP')
+sed -i "s/${PRIVATE_IP}/${PUBLIC_IP}/g" "$KUBECONFIG"
+
+# prepare a join command for workers to join the cluster
+set +o xtrace
+JOIN_CMD="$(gcloud compute ssh root@"$CP_NODE_NAME" --command "kubeadm token create --print-join-command")"
+JOIN_CMD="sudo $JOIN_CMD"
+
+# prepare worker startup script
+SETUP_STARTUP_WORKER=$TMP_DIR/setup-startup-worker.sh
+cat setup.sh >> "$SETUP_STARTUP_WORKER"
+cat startup-worker.sh >> "$SETUP_STARTUP_WORKER"
+chmod +x "$SETUP_STARTUP_WORKER"
+
+# start worker VMs and run the startup script that includes the join command
+for i in $(seq 1 "$NUM_WORKERS"); do
+    NODE="$CLUSTER_NAME-worker-$i"
+    echo "creating node $NODE"
+	gcloud compute instances create "$NODE" \
+	--metadata-from-file startup-script="$SETUP_STARTUP_WORKER" \
+    --metadata join-cmd="$JOIN_CMD" \
+	--image-family ubuntu-1804-lts \
+	--image-project ubuntu-os-cloud \
+	--machine-type $MACHINE_TYPE_WORKER \
+	--network "$CLUSTER_NAME" \
+	--labels=cluster="$CLUSTER_NAME" \
+	--tags kube-worker
+done
```

Comment:
does this scale well? 
should this be done in parallel?
don´t know if `gcloud compute instances create` allow to create batches of instance, but we can do it with bash too

Reply:
I don't think it will scale below small hundred of nodes. You should use Managed Instance Groups to create multiple instances. See how it's done in the kube-up.sh

Reply:
added the workers as managed groups.